cfme/tests/cloud/test_providers.py

```diff
@@ -22,20 +23,164 @@
 from cfme.fixtures.pytest_store import store
 from cfme.markers.env_markers.provider import ONE
 from cfme.rest.gen_data import arbitration_profiles as _arbitration_profiles
+from cfme.utils import appliance
+from cfme.utils import conf
+from cfme.utils import ssh
 from cfme.utils.appliance.implementations.ui import navigate_to
+from cfme.utils.conf import cfme_data
+from cfme.utils.conf import credentials
+from cfme.utils.generators import random_vm_name
+from cfme.utils.providers import get_crud_by_name
 from cfme.utils.providers import list_providers
 from cfme.utils.providers import ProviderFilter
 from cfme.utils.rest import assert_response
 from cfme.utils.update import update
+from cfme.utils.wait import wait_for_decorator
 
 pytestmark = [pytest.mark.provider([CloudProvider], scope="function")]
 
+# path, where powershell scripts are located and where cfme image is downloaded,
+# needed for test_create_azure_vm_from_azure_image
+CD_SPACE = 'cd /mnt/space && '
+
 
 @pytest.fixture(scope='function')
 def enable_regions(provider):
     enable_provider_regions(provider)
 
 
+@pytest.fixture(scope='function')
+def az_pwsh_vm(appliance, request):
+    """
+    azure_pwsh contains powershell and necessary scripts to upload vhd, create VM, get ip of the
+    resource and delete the VM.
+    Find the provider that contains that template.
+    TODO(anikifor): define the configuration of that VM as cloud-init file
+    """
+    providers_data = cfme_data.get("management_systems", {})
+
+    for provider, values in providers_data.items():
+        templates = values.get('templates')
+        if templates:
+            az_template = templates.get('powershell_vm')
+            if az_template:
+                provider_key = provider
+
+    if not provider_key:
```

Comment:
`provider_key` is set within the `if` blocks above, and if there was no matching provider found, this will trigger an `UnboundLocalError`.

I think you need a bit more efficient structure here to handle these cases. Take advantage of the tools in `cfme.utils.providers`.

```
filter = cfme.utils.providers.ProviderFilter(required_fields=[['templates', 'powershell_vm']])
provider_objects_list = cfme.utils.providers.list_providers(filters=[filter])
```

Reply:
@mshriver huge thanks! this is so much better.